runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java

```diff
@@ -83,62 +84,198 @@
-  }
+    @After
+    public void tearDown() {
+      pipelineResult = testPipeline.run();
+      pipelineResult.waitUntilFinish();
+      assertThat(pipelineResult.getState(), equalTo(PipelineResult.State.DONE));
+    }
 
-  private void waitForReady() {
-    try {
-      ManagedChannel channel = ManagedChannelBuilder.forTarget(expansionAddr).build();
-      ConnectivityState state = channel.getState(true);
-      for (int retry = 0; retry < 30 && state != ConnectivityState.READY; retry++) {
-        Thread.sleep(500);
-        state = channel.getState(true);
+    private void waitForReady() {
+      try {
+        ManagedChannel channel = ManagedChannelBuilder.forTarget(expansionAddr).build();
+        ConnectivityState state = channel.getState(true);
+        for (int retry = 0; retry < 30 && state != ConnectivityState.READY; retry++) {
+          Thread.sleep(500);
+          state = channel.getState(true);
+        }
+        channel.shutdownNow();
+      } catch (InterruptedException e) {
+        throw new RuntimeException("interrupted.");
       }
-      channel.shutdownNow();
-    } catch (InterruptedException e) {
-      throw new RuntimeException("interrupted.");
     }
-  }
 
+    private byte[] toStringPayloadBytes(String data) throws IOException {
+      Row configRow =
+          Row.withSchema(Schema.of(Field.of("data", FieldType.STRING)))
+              .withFieldValue("data", data)
+              .build();
+
+      ByteString.Output outputStream = ByteString.newOutput();
+      try {
+        RowCoder.of(configRow.getSchema()).encode(configRow, outputStream);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      ExternalTransforms.ExternalConfigurationPayload payload =
+          ExternalTransforms.ExternalConfigurationPayload.newBuilder()
+              .setSchema(SchemaTranslation.schemaToProto(configRow.getSchema(), false))
+              .setPayload(outputStream.toByteString())
+              .build();
+      return payload.toByteArray();
+    }
+
+    protected void singleInputOutputTest(Pipeline pipeline) throws IOException {
+      PCollection<String> col =
+          pipeline
+              .apply(Create.of("1", "2", "3"))
+              .apply(External.of(TEST_PREFIX_URN, toStringPayloadBytes("0"), expansionAddr));
+      PAssert.that(col).containsInAnyOrder("01", "02", "03");
+    }
+
+    protected void multiInputOutputWithSideInputTest(Pipeline pipeline) {
+      PCollection<String> main1 = pipeline.apply("createMain1", Create.of("a", "bb"));
+      PCollection<String> main2 = pipeline.apply("createMain2", Create.of("x", "yy", "zzz"));
+      PCollection<String> side = pipeline.apply("createSide", Create.of("s"));
+      PCollectionTuple pTuple =
+          PCollectionTuple.of("main1", main1)
+              .and("main2", main2)
+              .and("side", side)
+              .apply(External.of(TEST_MULTI_URN, new byte[] {}, expansionAddr).withMultiOutputs());
```

Comment:
Probably we should simplify the API so that users don't have to keep passing "new byte[] {}" for empty input ? (in a separate PR)

Reply:
Ack.